modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/frontend/css/variables/DefaultThemeScopedCSSVariablesProvider.java

```diff
@@ -0,0 +1,123 @@
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.token.definition.internal.frontend.css.variables;
+
+import com.liferay.frontend.css.variables.ScopedCSSVariables;
+import com.liferay.frontend.css.variables.ScopedCSSVariablesProvider;
+import com.liferay.frontend.token.definition.FrontendTokenDefinition;
+import com.liferay.frontend.token.definition.FrontendTokenDefinitionRegistry;
+import com.liferay.frontend.token.definition.parsed.FrontendToken;
+import com.liferay.frontend.token.definition.parsed.FrontendTokenCategory;
+import com.liferay.frontend.token.definition.parsed.FrontendTokenMapping;
+import com.liferay.frontend.token.definition.parsed.FrontendTokenSet;
+import com.liferay.frontend.token.definition.parsed.ParsedFrontendTokenDefinition;
+import com.liferay.portal.kernel.model.LayoutSet;
+import com.liferay.portal.kernel.service.LayoutSetLocalService;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.WebKeys;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Chema Balsas
+ */
+@Component(
+	property = "service.ranking:Integer=" + Integer.MAX_VALUE,
+	service = ScopedCSSVariablesProvider.class
+)
+public class DefaultThemeScopedCSSVariablesProvider
+	implements ScopedCSSVariablesProvider {
+
+	@Override
+	public Collection<ScopedCSSVariables> getScopedCSSVariablesCollection(
+		HttpServletRequest httpServletRequest) {
+
+		Map<String, String> cssVariables = new HashMap<>();
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		LayoutSet layoutSet = _layoutSetLocalService.fetchLayoutSet(
+			themeDisplay.getSiteGroupId(), false);
+
+		FrontendTokenDefinition frontendTokenDefinition =
+			_frontendTokenDefinitionRegistry.getFrontendTokenDefinition(
+				layoutSet.getThemeId());
+
+		ParsedFrontendTokenDefinition parsedFrontendTokenDefinition =
+			frontendTokenDefinition.getParsedFrontendTokenDefinition(
+				themeDisplay.getLocale());
+
+		Collection<FrontendTokenCategory> frontendTokenCategories =
+			parsedFrontendTokenDefinition.getFrontendTokenCategories();
+
+		for (FrontendTokenCategory frontendTokenCategory :
```

Comment:
Yeah, what I had in mind was, while parsing, to keep the entities separated, so you wouldn't need to loop through categories and sets to get to the tokens.

I think `Collection<FrontendToken> frontendTokenDefinition.getTokens(String locale)` would be very helpful.

Since we've basically parsed the whole thing, that should just mean keeping some additional structures to link entities outside of the root-type level that I guess we have here.

Just to clarify, I'd also keep this, so one can get the sets for a category and the tokens for a set... even the tokens for a category would be useful, but that would be a different API ;)

Reply:
So you propose to add helper methods at the root object for direct access, isn't it? I think it's useful, too :+1:

Reply:
yup!